### PR TITLE
[FLINK-16464][sql-client]result-mode tableau may shift when content contains Chinese String in SQL CLI

### PR DESCRIPTION
**Pull request by leonardBang**
## What is the purpose of the change

* This pull request  fix result-mode tableau may shift when content contains Chinese String. *

## Brief change log

  - *update file org/apache/flink/table/client/cli/CliTableauResultView.java*

## Verifying this change

 Add test in CliTableauResultViewTest.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): ( no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/CliTableauResultViewTest.java [hunk 1]
```diff
@@ -167,8 +177,9 @@ public void testBatchResult() {
 				"|   false | -2147483648 |  9223372036854775807 |                     (NULL) |    12345.06789 |    2020-03-01 18:39:14.123 |\n" +
 				"|    true |         100 | -9223372036854775808 |                 abcdefg111 |         (NULL) | 2020-03-01 18:39:14.123456 |\n" +
 				"|  (NULL) |          -1 |                   -1 | abcdefghijklmnopqrstuvwxyz |   -12345.06789 |                     (NULL) |\n" +
+				"|  (NULL) |          -1 |                   -1 |         这是一段中文 |   -12345.06789 |      2020-03-04 18:39:14.0 |\n"	+
```
Comment (JingsongLi) [inline flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/CliTableauResultViewTest.java]: Looks like we can not use UTF-8 length too...

Comment (leonardBang) [reply flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/CliTableauResultViewTest.java]: Hi, Jingsongli, 
Here's right,  the String  `'         这是一段中文 '`  has same length with `'abcdefghijklmnopqrstuvwxyz'` and shows well  when print in terminal. 
You can see the final view if you using  sql-client .

Comment (KurtYoung) [reply flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/CliTableauResultViewTest.java]: You sure about this, because it doesn't look right when I print out the string via System.out.println in IDE

Comment (KurtYoung) [reply flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/CliTableauResultViewTest.java]: It also doesn't seem right when I executed this test in my terminal by command line.

Comment (KurtYoung) [reply flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/CliTableauResultViewTest.java]: I think Jingsong is right, we can't use `String.getBytes().length` here. For example, the result of `"这是一段中文".getBytes().length()` is 18, but if we want the display correct, it should be 12.

Comment (leonardBang) [reply flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/CliTableauResultViewTest.java]: The problem is that UTF-8 char's display width is different with UTF-8 char's bytes length. `"这是一段中文".getBytes().length()` will return 18 because length of Chinese char is 3 in UTF-8 unicode，here we can use GBK unicode which Chinese char's length. 
But GBK is not common enough than UTF-8.

Comment (leonardBang) [reply flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/CliTableauResultViewTest.java]: Using GBK works well with Chinese, 
image


my concern is other language like Japanese/Korean have same probelm.

Comment (KurtYoung) [reply flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/CliTableauResultViewTest.java]: It seems this is a full width / half width character issue, but not necessarily an encoding problem.

File: flink-table/flink-sql-client/pom.xml [hunk 2]
```diff
@@ -109,6 +109,13 @@ under the License.
 			<version>3.9.0</version>
 		</dependency>
 
+		<!-- Tools to Unify display format for different languages -->
+		<dependency>
+			<groupId>com.ibm.icu</groupId>
```
Comment (KurtYoung) [inline flink-table/flink-sql-client/pom.xml]: We should also shade this jar into `flink-sql-client` jar?
If yes, then please also update the license files

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliTableauResultView.java [hunk 3]
```diff
@@ -60,9 +61,9 @@
 	private static final int NULL_COLUMN_WIDTH = CliStrings.NULL_COLUMN.length();
 	private static final int MAX_COLUMN_WIDTH = 30;
 	private static final int DEFAULT_COLUMN_WIDTH = 20;
+	private static final int COLUMN_TRUNCATED_FLAG_WIDTH = 3;
```
Comment (KurtYoung) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliTableauResultView.java]: use `COLUMN_TRUNCATED_FLAG.length()` instead?

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliTableauResultView.java [hunk 4]
```diff
@@ -276,13 +277,13 @@ private void printSingleRow(int[] colWidths, String[] cols) {
 		sb.append("|");
 		int idx = 0;
 		for (String col : cols) {
-			byte[] colBytes = getUTF8Bytes(col);
 			sb.append(" ");
-			if (colBytes.length <= colWidths[idx]) {
-				sb.append(StringUtils.repeat(' ', colWidths[idx] - colBytes.length));
+			int colWidth = getStringWidth(col);
```
Comment (KurtYoung) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliTableauResultView.java]: `int displayWidth = getStringDisplayWidth(col)`  would be more clear and accurate

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliTableauResultView.java [hunk 5]
```diff
@@ -276,13 +277,13 @@ private void printSingleRow(int[] colWidths, String[] cols) {
...
+			if (colWidth <= colWidths[idx]) {
+				sb.append(StringUtils.repeat(' ', colWidths[idx] - colWidth));
 				sb.append(col);
 			} else {
-				sb.append(subMaxString(col, colWidths[idx]));
+				sb.append(getFixedString(col, colWidths[idx]));
```
Comment (KurtYoung) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliTableauResultView.java]: `getFixedString` => `truncateString`?

Comment (KurtYoung) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliTableauResultView.java]: and second parameter can be replaced with `colWidths[idx] - COLUMN_TRUNCATED_FLAG_WIDTH` as `targetLength`

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliUtils.java [hunk 6]
```diff
@@ -111,4 +113,33 @@ public static void normalizeColumn(AttributedStringBuilder sb, String col, int m
 		}
 		return typesAsString;
 	}
+
+	public static int getStringWidth(String str) {
```
Comment (KurtYoung) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliUtils.java]: please add unit tests for both newly added methods.

File: flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/CliTableauResultViewTest.java [hunk 7]
```diff
@@ -156,19 +176,21 @@ public void testBatchResult() {
 
 		view.displayBatchResults();
 		view.close();
-
+		// note: about
```
Comment (KurtYoung) [inline flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/CliTableauResultViewTest.java]: what's this?

Comment (leonardBang) [reply flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/CliTableauResultViewTest.java]: unused note, I'll clear it.
Thanks for your review,  I’ll address soon.

File: flink-table/flink-sql-client/pom.xml [hunk 8]
```diff
@@ -7,7 +7,9 @@ regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at
+
```
Comment (KurtYoung) [inline flink-table/flink-sql-client/pom.xml]: revert this

Comment (leonardBang) [reply flink-table/flink-sql-client/pom.xml]: ok, I just want to unify the format because the license header is different with other pom.xml

File: flink-table/flink-sql-client/src/main/resources/META-INF/NOTICE [hunk 9]
```diff
@@ -9,3 +9,4 @@ See bundled license files for details.
 
 - org.jline:jline-terminal:3.9.0
 - org.jline:jline-reader:3.9.0
+- com.ibm.icu:icu4j:65.1
```
Comment (aljoscha) [inline flink-table/flink-sql-client/src/main/resources/META-INF/NOTICE:12]: @KurtYoung @leonardBang I don't think icu is a BSD license. Could you please fix this, or revert the changes.

Comment (KurtYoung) [reply flink-table/flink-sql-client/src/main/resources/META-INF/NOTICE:12]: I checked https://www.apache.org/legal/resolved.html, and ICU is in category A, isn't is sufficient?

Comment (aljoscha) [reply flink-table/flink-sql-client/src/main/resources/META-INF/NOTICE:12]: The license is ok to add, yes. But in this list it's listed under the `BSD` licenses. Icu is not licensed under a BSD license. See for example https://github.com/apache/flink/blob/4035f03a1de24b7e292466e9fa0a0de9cb10f358/flink-dist/src/main/resources/META-INF/NOTICE#L20, which has many different blocks for different licenses.

Comment (KurtYoung) [reply flink-table/flink-sql-client/src/main/resources/META-INF/NOTICE:12]: I see what you mean, thanks
